samples/net/sockets/tls_configurations/src/main.c

```diff
@@ -0,0 +1,231 @@
+
+	socket_fd = socket(addr.sin_family, SOCK_STREAM, IPPROTO_TLS_1_2);
+	if (socket_fd < 0) {
+		LOG_ERR("Failed to create TLS socket (%d)", errno);
+		return -errno;
+	}
+
+	sec_tag_t sec_tag_list[] = {
+#if defined(SOME_CERTIFICATE_SUPPORTED)
+		CA_CERTIFICATE_TAG,
+#endif
+#if defined(CONFIG_MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
+		PSK_TAG,
+#endif
+	};
+
+	ret = setsockopt(socket_fd, SOL_TLS, TLS_SEC_TAG_LIST,
+			sec_tag_list, sizeof(sec_tag_list));
+	if (ret < 0) {
+		LOG_ERR("Failed to set TLS_SEC_TAG_LIST option (%d)", errno);
+		return -errno;
+	}
+
+	/* HOSTNAME is only required for key exchanges that use a certificate. */
+#if defined(SOME_CERTIFICATE_SUPPORTED)
+	ret = setsockopt(socket_fd, SOL_TLS, TLS_HOSTNAME,
+			 "localhost", sizeof("localhost"));
+	if (ret < 0) {
+		LOG_ERR("Failed to set TLS_HOSTNAME option (%d)", errno);
+		return -errno;
+	}
+#endif
+
+	ret = connect(socket_fd, (struct sockaddr *) &addr, sizeof(addr));
+	if (ret < 0) {
+		LOG_ERR("Cannot connect to TCP remote (%d)", errno);
+		return -errno;
+	}
+
+	/* Prepare file descriptor for polling */
+	fds[0].fd = eventfd(0, 0);
+	fds[0].events = POLLIN;
+
+	fds[1].fd = socket_fd;
+	fds[1].events = POLLIN;
+
+	return ret;
+}
+
+void close_socket(void)
+{
+	if (socket_fd != INVALID_SOCKET) {
+		close(socket_fd);
+	}
+}
+
+static int setup_credentials(void)
+{
+	LOG_INF(APP_BANNER);
+	int err;
+
+#if defined(SOME_CERTIFICATE_SUPPORTED)
+	err = tls_credential_add(CA_CERTIFICATE_TAG,
+				TLS_CREDENTIAL_CA_CERTIFICATE,
+				certificate,
+				sizeof(certificate));
+	if (err < 0) {
+		LOG_ERR("Failed to register certificate: %d", err);
+		return err;
+	}
+#endif
+
+#if defined(CONFIG_MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
+	err = tls_credential_add(PSK_TAG,
+				TLS_CREDENTIAL_PSK,
+				psk,
+				sizeof(psk));
+	if (err < 0) {
+		LOG_ERR("Failed to register PSK: %d", err);
+	}
+	err = tls_credential_add(PSK_TAG,
+				TLS_CREDENTIAL_PSK_ID,
+				psk_id,
+				sizeof(psk_id) - 1);
+	if (err < 0) {
+		LOG_ERR("Failed to register PSK ID: %d", err);
+	}
+#endif
+
+	return 0;
+}
+
+int main(void)
+{
+	int ret;
+	size_t data_len;
+
+	setup_credentials();
+
+	k_thread_priority_set(k_current_get(), 0);
+
+	ret = create_socket();
+	if (ret < 0) {
+		LOG_ERR("Socket creation failed (%d)", ret);
+		goto exit;
+	}
+
+	memcpy(test_buf, TEST_STRING, sizeof(TEST_STRING));
+	data_len = sizeof(TEST_STRING) - 1;
+
+	for (int i = 0; i < 2; i++) {
+		LOG_DBG("Send: %s", test_buf);
+		ret = send(socket_fd, test_buf, data_len, 0);
+		if (ret < 0) {
+			LOG_ERR("Error sending test string (%d)", ret);
+			goto exit;
+		}
+
+		memset(test_buf, 0, sizeof(test_buf));
+
+		wait_for_event();
+
+		ret = recv(socket_fd, test_buf, sizeof(test_buf), 0);
+		if (ret <= 0) {
+			LOG_ERR("Error receiving data (%d)", ret);
+			goto exit;
+		}
+		/* The -1 here is because sizeof() accounts for "\0" whereas recv() does not. */
+		if (ret != (sizeof(TEST_STRING) - 1)) {
```

Comment:
That's risky because stream socket doesn't guarantee you won't receive data in chunks. If you expect `sizeof(TEST_STRING) - 1` bytes of data to receive, technically you should call `recv()` in a loop, until a total of `sizeof(TEST_STRING) -1` bytes is received, and only then compare the buffer content.

Reply:
What if I specify `MSG_WAITALL` as flag for `recv`? Shouldn't this solve the problem as well?

Reply:
That's also an option, maybe even simpler, just remember to replace `sizeof(test_buf)` with `sizeof(TEST_STRING) - 1` when calling `recv()`.